lib/cdo/rack/cookie_dcdo.rb

```diff
@@ -0,0 +1,34 @@
+require 'json'
+require 'request_store'
+require 'dynamic_config/dcdo'
+
+# This middleware enables the setting of DCDO via cookies for testing purposes.
+# See: 'dashboard/test/ui/features/cookie_dcdo.rb'
+module Rack
+  class CookieDCDO
+    KEY = 'DCDO'.freeze
+
+    def initialize(app)
+      @app = app
+    end
+
+    def call(env)
+      # Stores the cookie DCDO data per request.
+      RequestStore.store[KEY] = JSON.parse(Rack::Request.new(env).cookies[KEY] || '{}')
+
+      unless DCDO.instance_variable_get(:@_redefined)
+        # Redefines `DCDO#get` to return the cookie DCDO value if it exists.
+        DCDO.define_singleton_method(:get) do |key, *args|
+          RequestStore.store[KEY]&.key?(key) ? RequestStore.store[KEY][key] : super(key, *args)
+        end
+
+        DCDO.instance_variable_set(:@_redefined, true)
+      end
```

Comment:
I usually prefer to avoid metaprogramming, but in this case, I have three reasons for doing so:

1. Since this is intended for testing purposes only, I didn't want to modify the original method, as that would burden it with redundant checks in production and other environments.
2. This dynamic method redefinition will ensure that we override DCDO#get regardless of its origin, even if it isn't inherited from DynamicConfigBase where the original #get method is defined.
3. It's easier to understand the business logic when everything is kept together in one place.

Reply:
I suppose it does make sense to keep this test-only functionality isolated from the actual production implementation, but I'm still not a fan of the specific way we're going about it here. In particular, I really don't like that we've implemented the method redefinition somewhere that will be called repeatedly rather than as a onetime initialization.

What if we instead declare a module here in this file which can be prepended to the DCDO class in the initialize method? Similar to how we make modifications to existing classes in our initializers: [1](https://github.com/code-dot-org/code-dot-org/blob/c7f1bc053759520a7228b44727895088bb624148/dashboard/config/initializers/fast_localization.rb#L44), [2](https://github.com/code-dot-org/code-dot-org/blob/c7f1bc053759520a7228b44727895088bb624148/dashboard/config/initializers/mysql_avoid_utf8_alias_in_schema_dump.rb#L58), [3](https://github.com/code-dot-org/code-dot-org/blob/c7f1bc053759520a7228b44727895088bb624148/dashboard/config/initializers/trackable.rb#L21)

Reply:
That's how I implemented this at first:
https://github.com/code-dot-org/code-dot-org/blob/6f3f113491d7a65cd148abdabcfda73d9e77c83e/lib/cdo/rack/cookie_dcdo.rb#L9-L30

Reply:
I definitely prefer that!

Reply:
https://github.com/code-dot-org/code-dot-org/pull/59219/commits/b99ecbc8d24870a6f193fb9d1a1d3ed63fd244c7